ocaml/xapi/message_forwarding.ml

```diff
@@ -161,25 +161,29 @@ let log_exn_ignore ?(doc = "performing unknown operation") f x =
 
 
 (* Given an SR, return a PBD to use for some storage operation. *)
-(* In the case of SR.destroy we need to be able to forward the SR operation when all
-   PBDs are unplugged - this is the reason for the consider_unplugged_pbds optional
-   argument below. All other SR ops only consider plugged PBDs... *)
-let choose_pbd_for_sr ?(consider_unplugged_pbds=false) ~__context ~self () =
+(* In the case of SR.destroy and forget we need to be able to forward the SR
+   operation when all PBDs are unplugged. This is the reason for the
+   consider_unplugged_pbds optional argument below. All other SR ops only
+   consider plugged PBDs. *)
+let choose_pbds_for_sr ?(consider_unplugged_pbds=false) ~__context ~self () =
   let all_pbds = Db.SR.get_PBDs ~__context ~self in
-  let plugged_pbds = List.filter (fun pbd->Db.PBD.get_currently_attached ~__context ~self:pbd) all_pbds in
+  let plugged_pbds = List.filter (fun self -> Db.PBD.get_currently_attached ~__context ~self) all_pbds in
   let pbds_to_consider = if consider_unplugged_pbds then all_pbds else plugged_pbds in
-  if Helpers.is_sr_shared ~__context ~self then
-    let master = Helpers.get_master ~__context in
-    let master_pbds = Db.Host.get_PBDs ~__context ~self:master in
-    (* shared SR operations must happen on the master *)
-    match Listext.List.intersect pbds_to_consider master_pbds with
-    | pbd :: _ -> pbd (* ok, master plugged *)
-    | [] -> raise (Api_errors.Server_error(Api_errors.sr_no_pbds, [ Ref.string_of self ])) (* can't do op, master pbd not plugged *)
-  else
-    match pbds_to_consider with
-    | [] -> raise (Api_errors.Server_error(Api_errors.sr_no_pbds, [ Ref.string_of self ]))
-    | pdb :: _ -> pdb
+  let pbd_candidates =
+    if Helpers.is_sr_shared ~__context ~self then
+      let master = Helpers.get_master ~__context in
+      let master_pbds = Db.Host.get_PBDs ~__context ~self:master in
+      (* shared SR operations must happen on the master *)
+      Listext.intersect pbds_to_consider master_pbds
+    else
+      pbds_to_consider
+  in
+  match pbd_candidates with
+  | [] -> raise (Api_errors.Server_error(Api_errors.sr_no_pbds, [ Ref.string_of self ]))
+  | pdbs -> pdbs
 
+let choose_pbd_for_sr ?consider_unplugged_pbds ~__context ~self () =
+  List.hd @@ choose_pbds_for_sr ?consider_unplugged_pbds ~__context ~self ()
```

Comment:
Shouldn't this need the old wrapper for empty lists to raise `Api_errors.sr_no_pbds`? This would raise an internal error from List instead now.

Reply:
`List.hd` is generally to be avoided. A match to handle the `[]` case is better. What's the use of redefining the function anyway?

Reply:
`sr_no_pbds` is being raised in the case where the list is empty (the match is right above this function), making it impossible to see an internal error from list. Would you rather have the match to be repeated?